plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -819,6 +849,99 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
+    {
+        IcebergVacuumHandle vacuumHandle = (IcebergVacuumHandle) executeHandle.getProcedureHandle();
+
+        Table table = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        Map<String, String> properties = table.properties();
+        if (properties.containsKey(WRITE_LOCATION_PROVIDER_IMPL)) {
+            throw new TrinoException(NOT_SUPPORTED, "Table " + executeHandle.getSchemaTableName() + " specifies " + properties.get(WRITE_LOCATION_PROVIDER_IMPL) +
+                    " as a location provider. Writing to Iceberg tables with custom location provider is not supported.");
+        }
+        Duration retention = requireNonNull(vacuumHandle.getRetentionThreshold(), "retention is null");
+        Duration minRetention = requireNonNull(getVacuumMinRetention(session), "vacuumMinRetention is null");
+        checkArgument(retention.compareTo(
+                minRetention) >= 0,
+                "Retention specified (%s) is shorter than the minimum retention configured in the system (%s). " +
+                        "Minimum retention can be changed with %s configuration property or iceberg.%s session property",
+                retention,
+                minRetention,
+                IcebergConfig.VACUUM_MIN_RETENTION,
+                IcebergSessionProperties.VACUUM_MIN_RETENTION);
+
+        long expireTimestampMillis = System.currentTimeMillis() - retention.toMillis();
+        table.expireSnapshots()
+                .expireOlderThan(expireTimestampMillis)
+                .cleanExpiredFiles(true)
+                .commit();
+        deleteOrphanFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+        deleteOrphanMetadataFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+    }
+
+    private void deleteOrphanFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        Set<String> validDataFilePaths = Streams.stream(table.snapshots())
+                .map(Snapshot::snapshotId)
+                .flatMap(snapshotId -> Streams.stream(table.newScan().useSnapshot(snapshotId).planFiles()))
+                .map(fileScanTask -> URI.create(fileScanTask.file().path().toString()).getPath())
+                .collect(toImmutableSet());
+        scanAndDeleteNotValidFiles(table, session, schemaTableName, expireTimestamp, validDataFilePaths, "/data/");
+    }
+
+    private void deleteOrphanMetadataFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        ImmutableSet<String> manifests = Streams.stream(table.snapshots())
+                .flatMap(snapshot -> snapshot.allManifests().stream())
+                .map(ManifestFile::path)
+                .collect(toImmutableSet());
+        List<String> manifestLists = ReachableFileUtil.manifestListLocations(table);
+        List<String> otherMetadataFiles = Streams.concat(
```

Comment:
I am afraid of this one. If Iceberg in newer version adds extra metadata files we will start deleting those unless this code is updated. It will or will not be caught by tests.
What orphaned metadata files do we expect? Maybe we can skip this one altogether and only delete data files.
@alexjo2144 @findepi WDYT?

Reply:
Initially I had these skipped but then it was pointed out that Iceberg's spark procedure do checks metadata as well. I am willing to remove this ;)

Reply:
that's a concious decision and aligned with Iceberg on Spark behavior.
doing differently would mean we're not cleaning up garbage effectively.

if a future iceberg version changes, both this procedure and its spark equivalent will need to be updated.
cc @rdblue

Reply:
So from the offline discussion it looks like we want to keep it

Reply:
I think it's fine to delete extra metadata files if you're cleaning up.